Makefile

```
@@ -694,6 +694,7 @@ docker-build-all: $(addprefix docker-build-,$(ALL_ARCH)) ## Build docker images
 docker-build-%:
 	$(MAKE) ARCH=$* docker-build
 
+# Choice of images to build/push
 ALL_DOCKER_BUILD = core kubeadm-bootstrap kubeadm-control-plane docker-infrastructure in-memory-infrastructure test-extension clusterctl
```

```suggestion
ALL_DOCKER_BUILD ?= core kubeadm-bootstrap kubeadm-control-plane docker-infrastructure in-memory-infrastructure test-extension clusterctl
```
Q: Do we have to add this to make it overwritable via env var?

I assume not, just no idea why we usually use `?=` for stuff like this :)

---

I believe you do. Using `?=` allows the env var to take precedence if it's set whereas `=` just ignores the env var.

---

Hm, this would mean the PR wouldn't work in its current state, right?

---

I made this configurable since fork builds (as my company's) may not require building all components. For example, only the controllers may be required but not the test or CLI images.

---

I got this part. My question is, is it currently possible to overwrite this via an env var without changing the above to `?=`?

---

Sounds like we need ?=

> If you’d like a variable to be set to a value only if it’s not already set, then you can use the shorthand operator ‘?=’ instead of ‘=’. These two settings of the variable ‘FOO’ are identical (see [The origin Function](https://www.gnu.org/software/make/manual/make.html#Origin-Function)):
> https://www.gnu.org/software/make/manual/make.html#Setting

---

```bash
➜  export ALL_DOCKER_BUILD=core
➜ make docker-build
...
➜ /Library/Developer/CommandLineTools/usr/bin/make ARCH=arm64 docker-build-core docker-build-kubeadm-bootstrap docker-build-kubeadm-control-plane docker-build-docker-infrastructure docker-build-in-memory-infrastructure docker-build-test-extension docker-build-clusterctl

---

Ah sorry, I thought this was part of my changes already but it's a GitHub suggested change 🙈.

We definitely use this change with overridden `ALL_DOCKER_BUILD` already, and `make ALL_DOCKER_BUILD="core" ARCH=boom REGISTRY=myreg docker-push` on macOS works fine to build only the core image. I found it's because `make` will override variables if you set them on the command line. Therefore, `ALL_DOCKER_BUILD="core" ARCH=boom REGISTRY=myreg make docker-push`, i.e. using real environment variables only, doesn't work. I will quickly test again and switch multiple places to `?=`. Good find, thx!!

---

Ah thank you very much, that is good to know :)

---

Unless you don't want to make environment vs. command line parameters the same. Not sure if that is desired?!

---

I think it would be good if both works